core/deno.ts

```diff
@@ -7,7 +7,7 @@ export function getDenoDir(): string {
   // On Windows: %LOCALAPPDATA%/deno (%LOCALAPPDATA% = FOLDERID_LocalAppData)
   // On macOS: $HOME/Library/Caches/deno
   // If something fails, it falls back to $HOME/.deno
-  let denoDir = process.env.DENO_DIR;
+  let denoDir = process.env.VSCODE_CUSTOM_DENO_DIR || process.env.DENO_DIR;
```

Comment:
One more thing.

Because the functions in the core will be used in client/server/typescript-deno-plugin

So when the configuration changes, it cannot be changed immediately, and the extension needs to be restarted.

To fix this, We can watch configuration changes and then set environment variables

Reply:
Thanks @axetroy .

Yes, I have the line `this.startDenoLangaugeServer()` in the `sync` function at https://github.com/denoland/vscode_deno/pull/166/files#diff-bc3c0f2fe3aca9ddbe96347ffb00641eR409-R433

and the `sync` function is "registered" to watch for configuration https://github.com/denoland/vscode_deno/pull/166/files#diff-bc3c0f2fe3aca9ddbe96347ffb00641eR615-R617

so it will watch for configuration changes and restart the server already.

Reply:
> so it will watch for configuration changes and restart the server already.

Why doesn't the LSP watch configuration changes and then update the environment variables?

This should be a simple model.

Restarting the LSP is not necessary

Reply:
Do you mean to use some pub sub channel (e.g. LangeuageServer `sendNotification()` + `onNotification()`) to notify the running server of the environment changes so that I don't need to restart it?

p.s. I haven't read through the whole implementation of the language server yet. Some guidance to enlighten me is greatly appreciated. Thank you.

Reply:
Oh, I re-read what you said. Do you mean there is an even simpler way to watch for VSCode configuration changes inside the DenoLangaugeServer instance? So it does not even need to be triggered from the extension side on `watchConfiguration()` callback?

Reply:
Oh, I re-read what you said. Do you mean there is an even simpler way to watch for VSCode configuration changes inside the DenoLangaugeServer instance? So it does not even need to be triggered from the extension side on `watchConfiguration()` callback?

Reply:
@axetroy Thanks for the advice.

I have updated to listen to the config changes inside the language server.